### ISSUE TITLE
Support SVM output layer 

### ISSUE THREAD

**Issue by tmatas** [apache/incubator-mxnet#1810]
Hi all,
I found this really interesting article and I think it would be great if someone will implement this feature as an alternative to SoftmaxOutput!
http://arxiv.org/pdf/1306.0239.pdf

**Comment by jonasrla** [comment]
I am interested

**Comment by jonasrla** [comment]
As far as I understood, this article is suggesting the usage of SVM as alternative of SoftmaxOutput using L2-SVM as cost function. They claim the results are better in this setup.

I'll just start implement this SVM neuron based on what's already done with SoftMaxOutput.

**Comment by jonasrla** [comment]
Sorry, apparently this problem is bigger then I can actually handle, I'm having a hard time to even figure how to test.

**Comment by tmatas** [comment]
Wonderful!

**Comment by jonasrla** [comment]
I'm having a great deal of trouble trying figuring out where the SoftmaxOutput tests are @antinucleon, could help on this one?

**Comment by antinucleon** [comment]
Do you mean this unittest? https://github.com/dmlc/mxnet/blob/master/tests/python/unittest/test_operator.py#L203

Basically, we can test operator output with Numpy implementation.

**Comment by jonasrla** [comment]
Hey guy, this has nothing to do this issue, but the naming "check" trick to make nose ignore bothered me a little, so I fixed the problem, that is, I managed to change
from
`def check_softmax_with_ignore_label(xpu)`
to
`def test_softmax_with_ignore_label()`
and pass the nosetest.

Actually, I don't fully understand the issue, but to accomplish that I had to move the test_softmax function definition to the last position. I suspect it has something to do with mx.cpu(), but since I don't fully understand what CPU context object means, I can't help you any further.

@antinucleon @tqchen, Since no one ever complained about it, is this little fix any relevant to the project?

**Comment by jonasrla** [comment]
Hello,

Is there any parameter other then **margin size**, **regularization coefficient**? I'm letting the user decide between **L1-SVM** and **L2-SVM** too, so it makes optional 3 parameters besides data and labels.

I made a toy test based on the gtc_tutorial MNIST example and had some bad time because **NDArrayIter** expects by default some **"softmax_label"**. I worked around calling **mx.io._init_data** and setting the **default_name="svm_label"**, but I'm sure I shouldn't be calling a underlined function. It looks like a _bug_ ;)

Right now I'm having poor results with my toy test, 7.01%, I believe that's because I should be training a SVM model. I got a bit confused whether K on the derivatives equations are SVM weights or node weight, after this experiment I'm pretty convinced they are really for SVM.

I may need to create two **new Tensor operators**, **L1_SVM** and **L2_SVM**, just like **Softmax** did. For now I'll just release the cpu version, but I'll be glad to try the gpu, I just don't have any experience, yet.

**Comment by jonasrla** [comment]
Great news! I ran the MNIST toy test on both Softmax and SVM with the same configuration for both and got **97.93%** for **Softmax** against **SVM's 98.15 %** with L2-SVM cost function. It looks like the paper was right!

I'll format my repo for a pull request, if you guys my parameters suggestion

Will my toy test be any useful for you?

**Comment by sxjscience** [comment]
@jonasrla @tmatas It seems that we can combine Relu activation and MakeLoss to get the hinge loss. max(0, 1 - ty) can be written as relu(1 - ty). And we can put a mx.sym.MakeLoss after the activation.

**Comment by Trangle** [comment]
@tmatas How is gpu implementation making, thanks.

**Comment by jonasrla** [comment]
Believe it or not I'm back and finally I have a proper env with CUDA.
Just to be sure, right now it's about rewriting the SVMOutpt-inl.h this time using MakeLoss and Relu, right?
Does that solves the problem for cpu and gpu? If so, I'll do it and test for both cases.

**Comment by jonasrla** [comment]
@sxjscience, I was considering your idea of basing my solution on MakeLoss and Relu, but with the API as is it's unfeasible to make a solution using just operators. The sign varies depending if it's a expected answer or not and since we don't have one_hot_encode implemented on C++, it is not possible to write it functionally.

**Comment by yajiedesign** [comment]
cpu and mshadow gpu add now.
